### PR TITLE
crypto: add proptest against k256 with ffi wrapper impl

### PR DESCRIPTION
**Pull request by joyqvq**
https://github.com/MystenLabs/narwhal/issues/534

### CODE REVIEW

File: crypto/src/tests/secp256k1_tests.rs [hunk 1]
```diff
@@ -145,3 +145,29 @@ async fn signature_service() {
...
+        let key_pair = Secp256k1KeyPair::from(priv_key);
+        let signature = key_pair.sign(&message);
+        assert!(key_pair.public().verify(&message, &signature).is_ok());
+
+        // use k256 to construct private key with the same bytes and signs the same message
+        let priv_key_1 = k256::ecdsa::SigningKey::from_bytes(&r).map_err(|_e| signature::Error::new())?;
+        let pub_key_1 = priv_key_1.verifying_key();
+        let signature_1: k256::ecdsa::Signature = priv_key_1.sign(message);
+        assert!(pub_key_1.verify(message, &signature_1).is_ok());
+
```
Comment (huitseeker) [inline crypto/src/tests/secp256k1_tests.rs:169]: We might be able to check for the equality of the byte representation of the public keys here as well.

**Review by huitseeker — COMMENTED** [review]
This is great! Thanks @joyqvq !

Here are some of the tests I think could be both quick to write and instructive:
- given a byte string of the correct length, secp256k1 and k256 agree on whether you can create a private key from it (essentially testing the scalar representation checks on deserializaiton),
- given one of the three assets (privkey, pubkey, signature) generated using one of the libraries, you can "serialize" it (to a byte representation) and get it interpreted correctly by the other library (essentially a test of serialization)

File: crypto/src/tests/secp256k1_tests.rs [hunk 2]
```diff
@@ -145,3 +145,36 @@ async fn signature_service() {
...
+        // same byte string produces same private key
+        assert_eq!(key_pair.private().as_bytes(), priv_key_1.to_bytes().as_slice());
+
+        // same public key is created
+        assert_eq!(key_pair_copied.public().as_bytes(), pub_key_1.to_bytes().as_slice());
+
+        // same signatures produced from both implementations
+        assert_eq!(signature.as_ref(), ToFromBytes::as_bytes(&signature_1));
```
Comment (kchalkias) [inline crypto/src/tests/secp256k1_tests.rs:178]: Ooc,ECDSA is a randomised algorithm, how come the signature bytes produced for the same message are the same?
I would expect that the test should be a signature generated by implA should be verified by implB and vice versa.

Comment (joyqvq) [reply crypto/src/tests/secp256k1_tests.rs:178]: there is a function for sign with an extra nonce that produces randomized signature. do we want that? in that case the signature would be indeed different. 

```
    pub fn sign_ecdsa(&self, msg: &Message, sk: &SecretKey) -> Signature {
        self.sign_ecdsa_with_noncedata_pointer(msg, sk, ptr::null())
    }

    /// Constructs a signature for `msg` using the secret key `sk` and RFC6979 nonce
    /// and includes 32 bytes of noncedata in the nonce generation via inclusion in
    /// one of the hash operations during nonce generation. This is useful when multiple
    /// signatures are needed for the same Message and SecretKey while still using RFC6979.
    /// Requires a signing-capable context.
    pub fn sign_ecdsa_with_noncedata(
        &self,
        msg: &Message,
        sk: &SecretKey,
        noncedata: &[u8; 32],
    ) -> Signature {
        let noncedata_ptr = noncedata.as_ptr() as *const ffi::types::c_void;
        self.sign_ecdsa_with_noncedata_pointer(msg, sk, noncedata_ptr)
    }
```

similar function in k256:
```
    fn try_sign_with_rng(&self, rng: impl CryptoRng + RngCore, msg: &[u8]) -> Result<S, Error> {
        self.try_sign_digest_with_rng(rng, S::Digest::new_with_prefix(msg))
    }
```

Comment (joyqvq) [reply crypto/src/tests/secp256k1_tests.rs:178]: re: ecrecovery, there is a code change to use recoverable signature, i stacked the PR here: https://github.com/MystenLabs/narwhal/pull/564 (still tuning, WIP)

Comment (huitseeker) [reply crypto/src/tests/secp256k1_tests.rs:178]: Yay, folks are by default using RFC6979 🎉

**Review by kchalkias — APPROVED** [review]
Approving this, and we need a few extra upcoming PRs 
- clear api in our codebase for signing using a) randomized signatures b) deterministic RFC6979, default should be the randomized (just for consistency with most implementations out there and the main ECDSA standard).
- an extra test for randomized sigs (sign with implA, verify with implB and vice versa)
- comparison benchmark and test for ecrecover as well.
- ECDSA-NM (which is already in progress)

**Pull request merged**